Concurrency issue on the `executor.Run` function

I spotted a concurrency issue in the `executor.Run` func, which I was able to reproduce locally (it also works with the `-o` flag):
- Build/call a go program that will print to stderr and exit immediately after
- We call `harpoon` capture with the `-e` flag set (`sudo ./bin/harpoon capture -e -c -f main.main -- /tmp/test_app`)
- The `Run` function exits, and triggers the defer function which closes the channels.
- The gorutine might still process some messages and tries to send on a closed channel
- Kaboom: 
```
panic: send on closed channel

goroutine 36 [running]:
github.com/alegrey91/harpoon/internal/executor.Run.func2()
        /home/ale/Documents/harpoon/internal/executor/exec.go:42 +0x12a
created by github.com/alegrey91/harpoon/internal/executor.Run in goroutine 6
        /home/ale/Documents/harpoon/internal/executor/exec.go:29 +0x207
```

Example program
```
package main

import (
	"fmt"
	"os"
)

func main() {
	for i := range 10 {
		fmt.Fprintf(os.Stderr, "%v\n", i)
	}
}
```

After #60 we discussed a bit that probably we want to refactor the `Run` function, before diving deep into that, should we patch this bug in the current implementation?

--- a/README.md
+++ b/README.md
@@ -26,7 +26,7 @@ So, `main.doSomething` is the symbol of the function we want to trace using `har
 Then, let's run `harpoon` to extract the syscalls from the function `main.doSomething`:
 
 ```shell
-harpoon capture -f main.doSomething ./binary
+harpoon capture -f main.doSomething -- ./binary
 read
 sigaltstack
 gettid
@@ -57,6 +57,8 @@ You can easily download the latest release using the installation script:
 ```sh
 curl -s https://raw.githubusercontent.com/alegrey91/harpoon/main/install | sudo sh
 ```
+
+(If your current version is `<= v0.8.2`, remove it from `/usr/local/bin/` before installing the new one).
 
 ### Build
 

--- a/install
+++ b/install
@@ -23,5 +23,5 @@ printf "[download succeded]\n"
 
 # install binary
 chmod +x "$BINARY_RELEASE_NAME"
-mv "$BINARY_RELEASE_NAME" "/usr/local/bin/$BINARY_NAME"
+mv "$BINARY_RELEASE_NAME" "/usr/sbin/$BINARY_NAME"
 printf "[installation succeded]\n"

--- a/internal/ebpf/probesfacade/captor/capture.go
+++ b/internal/ebpf/probesfacade/captor/capture.go
@@ -220,6 +220,8 @@ func (ebpf *ebpfSetup) Capture(ctx context.Context, resultCh chan []uint32, erro
 	ebpf.pb.Poll(300)
 	// wait for args completion
 	wg.Wait()
+	close(cmdStdoutCh)
+	close(cmdStderrCh)
 	ebpf.pb.Stop()
 
 	// sending last remained syscalls

--- a/internal/executor/exec.go
+++ b/internal/executor/exec.go
@@ -12,8 +12,6 @@ import (
 // The cmdOutput argument is used to print the command output.
 func Run(cmd []string, cmdOutput, cmdError bool, wg *sync.WaitGroup, outputCh, errorCh chan<- string) {
 	defer func() {
-		close(outputCh)
-		close(errorCh)
 		wg.Done()
 	}()
 
@@ -26,8 +24,11 @@ func Run(cmd []string, cmdOutput, cmdError bool, wg *sync.WaitGroup, outputCh, e
 		return
 	}
 
-	go func() {
-		if cmdOutput {
+	var ioWg sync.WaitGroup
+	if cmdOutput {
+		ioWg.Add(1)
+		go func() {
+			defer ioWg.Done()
 			scanner := bufio.NewScanner(stdout)
 			for scanner.Scan() {
 				outputCh <- scanner.Text()
@@ -35,8 +36,12 @@ func Run(cmd []string, cmdOutput, cmdError bool, wg *sync.WaitGroup, outputCh, e
 			if err := scanner.Err(); err != nil {
 				outputCh <- fmt.Sprintf("error: %v", err)
 			}
-		}
-		if cmdError {
+		}()
+	}
+	if cmdError {
+		ioWg.Add(1)
+		go func() {
+			defer ioWg.Done()
 			scanner := bufio.NewScanner(stderr)
 			for scanner.Scan() {
 				errorCh <- scanner.Text()
@@ -44,9 +49,12 @@ func Run(cmd []string, cmdOutput, cmdError bool, wg *sync.WaitGroup, outputCh, e
 			if err := scanner.Err(); err != nil {
 				errorCh <- fmt.Sprintf("error: %v", err)
 			}
-		}
-	}()
+		}()
+	}
 
+	// wait for stdout/stderr scans to be completed
+	ioWg.Wait()
+	// wait for the executed command to be completed
 	command.Wait()
 }
 

--- a/tests/integration.txtar
+++ b/tests/integration.txtar
@@ -42,8 +42,18 @@ exec chmod +x bin/example-app
 exec ls -la bin/example-app
 exec bin/example-app coin
 
+exec harpoon capture -f main.main -- ./bin/example-app coin
+stdout 'write'
+
 exec harpoon capture -f main.main -c -- ./bin/example-app coin
 stdout 'stdout: \[flip coin\]'
+
+exec harpoon capture -f main.main -e -- ./bin/example-app streams
+stdout 'stderr: 0'
+
+exec harpoon capture -f main.main -c -e -- ./bin/example-app streams
+stdout 'stdout: 0'
+stdout 'stderr: 0'
 
 exec harpoon capture -f main.main -- ./bin/example-app coin
 stdout 'write'

--- a/tests/testcases/example-app/cmd/streams.go
+++ b/tests/testcases/example-app/cmd/streams.go
@@ -1,0 +1,39 @@
+/*
+Copyright © 2023 NAME HERE <EMAIL ADDRESS>
+*/
+package cmd
+
+import (
+	"github.com/spf13/cobra"
+
+	"github.com/alegrey91/seccomp-test-coverage/pkg/streams"
+)
+
+// streamsCmd represents the streams command
+var streamsCmd = &cobra.Command{
+	Use:   "streams",
+	Short: "A brief description of your command",
+	Long: `A longer description that spans multiple lines and likely contains examples
+and usage of using your command. For example:
+
+Cobra is a CLI library for Go that empowers applications.
+This application is a tool to generate the needed files
+to quickly create a Cobra application.`,
+	Run: func(cmd *cobra.Command, args []string) {
+		streams.PrintOutErr()
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(streamsCmd)
+
+	// Here you will define your flags and configuration settings.
+
+	// Cobra supports Persistent Flags which will work for this command
+	// and all subcommands, e.g.:
+	// streamsCmd.PersistentFlags().String("foo", "", "A help for foo")
+
+	// Cobra supports local flags which will only run when this command
+	// is called directly, e.g.:
+	// streamsCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
+}

--- a/tests/testcases/example-app/pkg/streams/streams.go
+++ b/tests/testcases/example-app/pkg/streams/streams.go
@@ -1,0 +1,13 @@
+package streams
+
+import (
+	"fmt"
+	"os"
+)
+
+func PrintOutErr() {
+	for i := 0; i < 10; i++ {
+		fmt.Fprintf(os.Stderr, "%v\n", i)
+		fmt.Fprintf(os.Stdout, "%v\n", i)
+	}
+}

